exporters/provider_common/openshift.py

```diff
@@ -19,3 +69,234 @@ def convert_datetime(dt: Union[str, datetime]) -> datetime:
+                    return {owner_ref.uid: replica}
+        except ResourceNotFoundError:
+            logging.debug(
+                "API Object not found for version: %s object: %s",
+                owner_ref.apiVersion,
+                owner_ref.uid,
+            )
+    return {}
+
+
+def filter_pods_by_replica_uid(
+    pods_list: List[ResourceField],
+) -> Dict[str, ResourceField]:
+    """
+    Filters out the given list of Pod objects to create a dictionary with ReplicaSet or
+    ReplicationController UIDs as keys.
+
+    In OpenShift, one Pod may have multiple ownerReferences. This function filters the list of
+    Pod objects to identify unique ReplicaSet or ReplicationController objects based on their UID,
+    which is stored in the ownerReferences attribute of the Pod object.
+
+    Since Pods within a ReplicaSet are homogeneous and identical replicas,
+    we can filter out unique ReplicaSet objects even if multiple Pods are deployed.
+
+    Args:
+        pods_list (List[ResourceField]): A list of Pod objects.
+
+    Returns:
+        Dict[str, ResourceField]
+                A dictionary with ReplicaSet or ReplicationController UIDs as keys and Pod objects as values.
+    """
+    uids = set()
+    pods = {}
+
+    for pod in pods_list:
+        for owner_reference in pod.metadata.ownerReferences or []:
+            replica_uid = owner_reference.uid
+            if replica_uid in uids:
+                continue
+            uids.add(replica_uid)
+            pods[replica_uid] = pod
+
+    return pods
+
+
+def get_and_log_namespaces(
+    client: DynamicClient, namespaces: set[str], prod_label: str
+) -> set[str]:
+    """
+    Get the set of namespaces to watch, and log what they are.
+    They will be either:
+    1. The namespaces explicitly specified
+    2. The namespaces matched by PROD_LABEL
+    3. If neither namespaces nor the PROD_LABEL is given, then implicitly matches all namespaces.
+    """
+    if namespaces:
+        logging.info("Watching namespaces %s", namespaces)
+        return namespaces
+
+    if prod_label:
+        logging.info(
+            "No namespaces specified, watching all namespaces with given PROD_LABEL (%s)",
+            prod_label,
+        )
+        query_args = dict(label_selector=prod_label)
+    else:
+        logging.info(
+            "No namespaces specified and no PROD_LABEL given, watching all namespaces."
+        )
+        query_args = dict()
+
+    all_namespaces = client.resources.get(api_version="v1", kind="Namespace")
+    namespaces = {ns.metadata.name for ns in all_namespaces.get(**query_args).items}
+    logging.info("Watching namespaces %s", namespaces)
+    if not namespaces:
+        logging.warning(
+            "No NAMESPACES given and PROD_LABEL did not return any matching namespaces."
+        )
+    return namespaces
+
+
+def _extract_image_sha(image_url_or_id: str) -> Optional[Tuple[str, str, str]]:
```

Comment:
Here I think is over complicated (and the name/description doesn't seen right for me: gets registry and image name as well)

why don't return string or None instead of Tuple?

Reply:
It's just to make the function more generic. We could have only (sha256, full URI), but really I want to make sure we are also able to get just the image name. In situations where we use internal registry then using `{internal-registry}image_name@sha_value` may seem like a valid scenario and will improve dramatically calls skipping external registry queries.

Reply:
Done.

Renamed to `_parse_container_image_uri`, changed return value to be Union of values or Tuple of None's with the following description:

```
def _parse_container_image_uri(image_uri: str) -> Union[Tuple[str, str, str], Tuple[None, None, None]]:
    """
    Parses the container image URI and extracts image registry, image name and image SHA256 value.

    Args:
        image_uri[str]: Container image URI. 
            Expected is an URI with registry URI and SHA256 value.

    Returns:
        Tuple[str, str, str]
            Parsed container URI into Tuple of registry URI, image name and SHA256 value.
            If any of the expected values is not found it then returns (None, None, None).
    """
```